### ISSUE TITLE
pull Yubikey locks the database

### ISSUE THREAD

**Issue by Wanderingidea** [keepassxreboot/keepassxc#3616]
Security feature:
If using a Yubikey for challenge/response, pulling the Yubikey locks the database.
This probably means a polling timer to see if the Yubikey is still present.

**Comment by phoerious** [comment]
Polling the YubiKey is quite slow and I wouldn't be surprised if it depleted your laptop battery rather quickly. If something like this were to be implemented, we'd instead need to subscribe to USB device changes. However, that means meddling with platform-specific APIs, which is the reason we are polling in the first place (and therefore you need to hit the "Refresh" button if you insert the key after opening the KeePassXC window).

**Comment by Wanderingidea** [comment]
That is why it is a good idea to make it an option.

**Comment by phoerious** [comment]
No. It would only make it mandatory to make it an option. It doesn't mean the option itself is a good idea in the first place.

**Comment by Wanderingidea** [comment]
Of course, but I still think it is a good idea. Not everyone uses a laptop and it is not certain the laptop battery depletes too quickly.

**Comment by droidmonkey** [comment]
The ideal way to deal with this situation is to subscribe to usb events and poll for the Yubikey on a disconnect event.

**Comment by atticus-sullivan** [comment]
Possible current Workaround for the linux users: Write an udev rule like described at https://wiki.archlinux.org/title/YubiKey#Tips_and_tricks and send a notification to close via (q)dbus (https://github.com/keepassxreboot/keepassxc/wiki/Using-DBus-with-KeePassXC).

Nonetheless: A native feature for this would be nice

**Comment by GrimzEcho** [comment]
It sounds like this already would be gated behind an option if ever implemented, but just in case...

Using Yubikeys over NFC is also supported now, so something that listens for USB events might not be enough. That said, for some NFC use cases this behavior wouldn't be desirable so it would have to be behind an option.

**Comment by acolomb** [comment]
As there is proper USB notification support in KeePassXC now, could this be considered again, please?

I suppose the listening part is implemented now, based on the new user experience with the "Use hardware key" option automatically showing / hiding on a locked database's tab.

**Comment by droidmonkey** [comment]
This feature would really suck if there was a change that needed to be saved since you'll have to plug your yubikey back in to actually save and lock the database. I am not in favor of implementing this, or if implemented it only locks if there are no changes to save.

**Comment by acolomb** [comment]
First of all, I meant it to be a purely opt-in mechanism, just like the existing two "Lock databases when / after..." options.

As to requiring the key in case of unsaved changes: I'm in no position to declare my usage is typical. But for me, by far more often I do *not* have unsaved changes when locking the database than the occasions where I do.

Now let's compare the UX in that case only. I pulled my key right after unlocking. The configured timeout expires and a dialog pops up asking me whether to save. Assuming I notice it, that's when I need to insert the key for saving. Same with closing the lid or automatic session lock, except there I don't even get a chance to see the dialog before opening it again. Yet both options are provided.

In contrast, when I pull the key from USB, I still have it in my hand. When the dialog asks the question in that moment, it's the most convenient time to just plug it back in and confirm saving. Much easier than the other two occasions. So I still think it will be useful in most cases and less annoying than other implemented features in the remaining cases.

**Comment by droidmonkey** [comment]
Fair point, and yes it will be opt in

**Comment by pfzetto** [comment]
> Possible current Workaround for the linux users: Write an udev rule like described at https://wiki.archlinux.org/title/YubiKey#Tips_and_tricks and send a notification to close via (q)dbus (https://github.com/keepassxreboot/keepassxc/wiki/Using-DBus-with-KeePassXC).
> 
> Nonetheless: A native feature for this would be nice

I've tested this method and there is a case where it doesn't work as expected.
If the database has pending changes, KeepassXC waits for the user to present the Yubikey and stays unlocked until then.
I think a feature like #9089 would be needed to support save and lock when the Yubikey is removed.
